### PR TITLE
feat: [IDP-150] Infra Setup Pod Health Check

### PR DESCRIPTION
**Pull request by vigneswara-propelo**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-ALL: `trigger tiAll`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/44101)

### CODE REVIEW

**Review by vikyathharekal — COMMENTED** [review]
> Reviewed 6 of 7 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44101)* status: 7 file(s) to review. 0 of 2 approvals obtained. 

___
*[`idp-service/service/src/main/java/io/harness/idp/status/enums/Status.java` line 12 at r2](https://reviewable.io/reviews/harness/harness-core/44101#-NOjQmiLDPeOt9KHmQWi:-NOjQmiLDPeOt9KHmQWj:b-v2cwpn) ([raw file](https://github.com/harness/harness-core/blob/1cad2c80647feacbee17e3e9b569c288211be635/idp-service/service/src/main/java/io/harness/idp/status/enums/Status.java#L12)):*
> ```Java
> //  Wiki page specifying different statuses
> //  https://harness.atlassian.net/wiki/spaces/IDP/pages/21301300949/IDP+App+Status+logic+-+Tech+Spec#Different-Statuses
> public enum Status { COMPLETED, RUNNING, PENDING, FAILED, NOT_EXIST }
> ```

nit: NOT_FOUND sounds better to me



_Code quote:_
```Java
NOT_EXIST
```
___
*[`idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java` line 33 at r2](https://reviewable.io/reviews/harness/harness-core/44101#-NOjPRXo2D2fbTy9ZJV0:-NOjPRXo2D2fbTy9ZJV1:bkxuthp) ([raw file](https://github.com/harness/harness-core/blob/1cad2c80647feacbee17e3e9b569c288211be635/idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java#L33)):*
> ```Java
> 
> @Slf4j
> public class PodHealthCheck implements HealthCheck {
> ```

Can this whole class go into the idp-service -> modules -> k8s -> health?
___
*[`idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java` line 48 at r2](https://reviewable.io/reviews/harness/harness-core/44101#-NOjNvk_5qvIYHITGtnm:-NOjNvk_5qvIYHITGtnn:bmz2wwt) ([raw file](https://github.com/harness/harness-core/blob/1cad2c80647feacbee17e3e9b569c288211be635/idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java#L48)):*
> ```Java
>       // TODO: Implement logic for pod restart scenarios
>       V1PodList podList =
>           api.listNamespacedPod(namespace, null, null, null, null, LABEL_SELECTOR, null, null, null, null, false);
> ```

Can we add label filters to just get the pod we need? For now, the namespace could just have one pod, but better to get the exact pod we need.
___
*[`idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java` line 66 at r2](https://reviewable.io/reviews/harness/harness-core/44101#-NOjN57l25R6ZLzLAjH4:-NOjN57l25R6ZLzLAjH5:b-nt2xwk) ([raw file](https://github.com/harness/harness-core/blob/1cad2c80647feacbee17e3e9b569c288211be635/idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java#L66)):*
> ```Java
>     } catch (Exception e) {
>       String err = "Could not check for pod status, " + e.getMessage();
>       log.error(err);
> ```

Could we log the whole stack trace? It would be beneficial while debugging prod issues later.
The status Info could still have just the error message as you have done.
___
*[`idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java` line 93 at r2](https://reviewable.io/reviews/harness/harness-core/44101#-NOjP7EHDOeXqntsy9qa:-NOjP7EHDOeXqntsy9qb:bnyr2ma) ([raw file](https://github.com/harness/harness-core/blob/1cad2c80647feacbee17e3e9b569c288211be635/idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java#L93)):*
> ```Java
>   }
> 
>   private String getPodMessage(V1Pod pod) {
> ```

What kind of messages do we actually get? Do you have any examples? Just to understand 
___
*[`idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java` line 104 at r2](https://reviewable.io/reviews/harness/harness-core/44101#-NOjOkJw3AQAC488WZuY:-NOjOkJw3AQAC488WZuZ:b-xwzzfh) ([raw file](https://github.com/harness/harness-core/blob/1cad2c80647feacbee17e3e9b569c288211be635/idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java#L104)):*
> ```Java
>               .map(containerStatus -> containerStatus.getState().getWaiting().getMessage())
>               .collect(Collectors.toList());
>       msg = String.join(". ", containerMsgs);
> ```

Can we define a CONSTANT for this separator string? 



_Code quote:_
```Java
". "
```
___
*[`idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java` line 112 at r2](https://reviewable.io/reviews/harness/harness-core/44101#-NOjOtOFDnhbB2cRasmu:-NOjOtOFDnhbB2cRasmv:b7uorkh) ([raw file](https://github.com/harness/harness-core/blob/1cad2c80647feacbee17e3e9b569c288211be635/idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java#L112)):*
> ```Java
>                                           .map(V1PodCondition::getMessage)
>                                           .collect(Collectors.toList());
>       msg = String.join(". ", podConditionMsgs);
> ```

We could use `.collect(Collectors.joining(","))` instead
___
*[`idp-service/service/src/main/java/io/harness/idp/status/service/StatusInfoServiceImpl.java` line 36 at r2](https://reviewable.io/reviews/harness/harness-core/44101#-NOjQS6_-i_jjcqbin42:-NOjQS6_-i_jjcqbin43:b-7s6h1k) ([raw file](https://github.com/harness/harness-core/blob/1cad2c80647feacbee17e3e9b569c288211be635/idp-service/service/src/main/java/io/harness/idp/status/service/StatusInfoServiceImpl.java#L36)):*
> ```Java
>       return healthCheck.getCurrentStatus(accountIdentifier);
>     }
>     Optional<StatusInfoEntity> statusEntity =
> ```

Where/When do we save the status for the ONBOARDING type?

File: idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import javax.inject.Inject;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.collections.CollectionUtils;
+
+@Slf4j
+public class PodHealthCheck implements HealthCheck {
+  @Inject private KubernetesHelperService kubernetesHelperService;
+  @Inject private NamespaceService namespaceService;
+
+  private static final String LABEL_SELECTOR = "app=idp-ui";
```
Comment (sandeepa-kv-jois) [inline idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java:37]: For now, we have one. There could be more than one type in the future if we decide to have sidecar for any additional functionality and we may need to get health status for all.  It's better to have it as list.

Comment (sandeepa-kv-jois) [reply idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java:37]: Also, it's good to read any values from the manifest(as env variable) than having them hard-coding wherein any change to this value would require a hotfix.

**Review by vigneswara-propelo — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44101)* status: 7 file(s) to review. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java` line 33 at r2](https://reviewable.io/reviews/harness/harness-core/44101#-NOjPRXo2D2fbTy9ZJV0:-NOlubnN-u_TD3fXc4xw:b-pgthzj) ([raw file](https://github.com/harness/harness-core/blob/1cad2c80647feacbee17e3e9b569c288211be635/idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java#L33)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Can this whole class go into the idp-service -> modules -> k8s -> health?
</blockquote></details>

it will create cyclic dependency if we want to access DB for restart scenarios etc
___
*[`idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java` line 37 at r2](https://reviewable.io/reviews/harness/harness-core/44101#-NOiGbhwoCxlzE9JFphO-r2-37:-NOlyd3L6YWKrySLqLq8:bqp84gu) ([raw file](https://github.com/harness/harness-core/blob/1cad2c80647feacbee17e3e9b569c288211be635/idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java#L37)):*
<details><summary><i>Previously, sandeepa-kv-jois wrote…</i></summary><blockquote>

Also, it's good to read any values from the manifest(as env variable) than having them hard-coding wherein any change to this value would require a hotfix.
</blockquote></details>

listNamespacedPod API accepts only string type for labelSelector. we can have something like this "app in (idp-ui, idp-backend)" for matching multiple labels 
___
*[`idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java` line 48 at r2](https://reviewable.io/reviews/harness/harness-core/44101#-NOjNvk_5qvIYHITGtnm:-NOlv0EH0NfXL0Ei_R6V:b-2fw3dl) ([raw file](https://github.com/harness/harness-core/blob/1cad2c80647feacbee17e3e9b569c288211be635/idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java#L48)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Can we add label filters to just get the pod we need? For now, the namespace could just have one pod, but better to get the exact pod we need.
</blockquote></details>

LABEL_SELECTOR constant will do the filtering here
___
*[`idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java` line 93 at r2](https://reviewable.io/reviews/harness/harness-core/44101#-NOjP7EHDOeXqntsy9qa:-NOlxpTDCPQ30hOY4Z8U:bjq052j) ([raw file](https://github.com/harness/harness-core/blob/1cad2c80647feacbee17e3e9b569c288211be635/idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java#L93)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

What kind of messages do we actually get? Do you have any examples? Just to understand 
</blockquote></details>

One such message would be like below if the image is invalid



_Code snippet:_
```Java
rpc error: code = Unknown desc = Error response from daemon: pull access denied for vigneswara1998/python-demo-harness-qa1, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
```

**Review by vigneswara-propelo — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44101)* status: 7 file(s) to review. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/service/src/main/java/io/harness/idp/status/service/StatusInfoServiceImpl.java` line 36 at r2](https://reviewable.io/reviews/harness/harness-core/44101#-NOjQS6_-i_jjcqbin42:-NOm0JlY7HeaS3IEJryo:b-gamrof) ([raw file](https://github.com/harness/harness-core/blob/1cad2c80647feacbee17e3e9b569c288211be635/idp-service/service/src/main/java/io/harness/idp/status/service/StatusInfoServiceImpl.java#L36)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Where/When do we save the status for the ONBOARDING type?
</blockquote></details>

we will be saving onboarding status once the onboarding workflow is done

**Review by vigneswara-propelo — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44101)* status: 7 file(s) to review. 8. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/service/src/main/java/io/harness/idp/status/enums/Status.java` line 12 at r2](https://reviewable.io/reviews/harness/harness-core/44101#-NOjQmiLDPeOt9KHmQWi:-NOmj3jU9q2WlwlBeYos:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/1cad2c80647feacbee17e3e9b569c288211be635/idp-service/service/src/main/java/io/harness/idp/status/enums/Status.java#L12)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

nit: NOT_FOUND sounds better to me
</blockquote></details>

Done
___
*[`idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java` line 37 at r2](https://reviewable.io/reviews/harness/harness-core/44101#-NOiGbhwoCxlzE9JFphO-r2-37:-NOmi5g003CtjiMb7zkB:bh9p83k) ([raw file](https://github.com/harness/harness-core/blob/1cad2c80647feacbee17e3e9b569c288211be635/idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java#L37)):*
<details><summary><i>Previously, vigneswara-propelo (Vigneswara) wrote…</i></summary><blockquote>

listNamespacedPod API accepts only string type for labelSelector. we can have something like this "app in (idp-ui, idp-backend)" for matching multiple labels 
</blockquote></details>

yeah totally agree, will create a separate PR to handle fetching LABEL_SELECTOR from manifest file. Created a child task IDP-193 for the same
___
*[`idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java` line 66 at r2](https://reviewable.io/reviews/harness/harness-core/44101#-NOjN57l25R6ZLzLAjH4:-NOmj16h-BUoi55X9B26:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/1cad2c80647feacbee17e3e9b569c288211be635/idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java#L66)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Could we log the whole stack trace? It would be beneficial while debugging prod issues later.
The status Info could still have just the error message as you have done.
</blockquote></details>

Done.
___
*[`idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java` line 104 at r2](https://reviewable.io/reviews/harness/harness-core/44101#-NOjOkJw3AQAC488WZuY:-NOmj0Cv9z81-ZREktff:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/1cad2c80647feacbee17e3e9b569c288211be635/idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java#L104)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Can we define a CONSTANT for this separator string? 
</blockquote></details>

Done.
___
*[`idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java` line 112 at r2](https://reviewable.io/reviews/harness/harness-core/44101#-NOjOtOFDnhbB2cRasmu:-NOmj-OB4gJmcLZ8aKWi:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/1cad2c80647feacbee17e3e9b569c288211be635/idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java#L112)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

We could use `.collect(Collectors.joining(","))` instead
</blockquote></details>

Done.

File: idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java (hunk 1 above)
Comment (vigneswara-propelo) [reply idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java:36]: listNamespacedPod API accepts only string type for labelSelector. we can have something like this "app in (idp-ui, idp-backend)" for matching multiple labels

yeah totally agree, will create a separate PR to handle fetching LABEL_SELECTOR from manifest file. Created a child task IDP-193 for the same

**Review by sandeepa-kv-jois — APPROVED** [review]
> Reviewed 3 of 7 files at r1, 1 of 2 files at r3, 1 of 1 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44101)* status: 4 file(s) to review. 7. unresolved discussions. 1 of 2 approvals obtained.

**Review by deepharness — CHANGES_REQUESTED** [review]
> Reviewed 4 of 7 files at r1, 2 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44101)* status: 2 file(s) to review. 1 change requests. 7. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`idp-service/modules/k8s/src/main/java/io/harness/k8s/client/K8sApiClient.java` line 121 at r5](https://reviewable.io/reviews/harness/harness-core/44101#-NOnjMVO7paOd1iO4Mdg:-NOnjMVP90JxW4Lf4_Me:balu24a) ([raw file](https://github.com/harness/harness-core/blob/c77d39954a876f3aee9312c19b3d45fa28149e86/idp-service/modules/k8s/src/main/java/io/harness/k8s/client/K8sApiClient.java#L121)):*
> ```Java
> 
>   public static KubernetesConfig getKubernetesConfig(String namespace) {
>     if (StringUtils.isBlank(namespace)) {
> ```

do we need this namespace only for null check ?

**Review by vigneswara-propelo — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44101)* status: 2 file(s) to review. 1 change requests. 7. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`idp-service/modules/k8s/src/main/java/io/harness/k8s/client/K8sApiClient.java` line 121 at r5](https://reviewable.io/reviews/harness/harness-core/44101#-NOnkjlk--OiJVw24JYn:-NOnkjll9vVXEIKKuCVf:bdudyrd) ([raw file](https://github.com/harness/harness-core/blob/c77d39954a876f3aee9312c19b3d45fa28149e86/idp-service/modules/k8s/src/main/java/io/harness/k8s/client/K8sApiClient.java#L121)):*
> ```Java
> 
>   public static KubernetesConfig getKubernetesConfig(String namespace) {
>     if (StringUtils.isBlank(namespace)) {
> ```

yes, its not required. Tried to remove namespace from this method, but there are some conflicts with the latest code. Will clean this up in a separate PR

**Review by deepharness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44101)* status: 2 file(s) to review. 7. unresolved discussions.

**Review by vikyathharekal — COMMENTED** [review]
> Reviewed 1 of 2 files at r3, 1 of 1 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44101)* status: :shipit: complete! ✓

___
*[`idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java` line 33 at r2](https://reviewable.io/reviews/harness/harness-core/44101#-NOjPRXo2D2fbTy9ZJV0:-NOqeyCw6zcW06F65laB:b-by8bo3) ([raw file](https://github.com/harness/harness-core/blob/1cad2c80647feacbee17e3e9b569c288211be635/idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java#L33)):*
<details><summary><i>Previously, vigneswara-propelo (Vigneswara) wrote…</i></summary><blockquote>

it will create cyclic dependency if we want to access DB for restart scenarios etc
</blockquote></details>

My thinking was - the DB accessing code could be in the status module and the k8s.health module could return whatever data is needed by the status module.
Let's connect for 5 mins to understand if this would work or not.

**Review by vikyathharekal — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44101)* status: :shipit: complete! ✓

___
*[`idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java` line 33 at r2](https://reviewable.io/reviews/harness/harness-core/44101#-NOjPRXo2D2fbTy9ZJV0:-NOrF_JA9TbT1QLiNlFR:b52wyfv) ([raw file](https://github.com/harness/harness-core/blob/1cad2c80647feacbee17e3e9b569c288211be635/idp-service/service/src/main/java/io/harness/idp/status/k8s/PodHealthCheck.java#L33)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

My thinking was - the DB accessing code could be in the status module and the k8s.health module could return whatever data is needed by the status module.
Let's connect for 5 mins to understand if this would work or not.
</blockquote></details>

Discussed offline. We could revisit this comment when we work on the DB part later

**Pull request merged**